src/lib/array.jl

```diff
@@ -80,26 +80,20 @@ _droplike(dy::Union{LinearAlgebra.Adjoint, LinearAlgebra.Transpose}, dxv::Abstra
 
 for f in [push!, pop!, pushfirst!, popfirst!]
   @eval @adjoint! $f(xs, x...) =
-    push!(xs, x...), _ -> error("Mutating arrays is not supported")
+    $f(xs, x...), _ -> error("Mutating arrays is not supported")
 end
 
-# This is kind of bad, but at least we don't materialize the whole
-# array. Prefer to use `Buffer`
-# function _pullback(cx::Context, ::typeof(push!), xs::AbstractVector{<:AbstractArray}, x::AbstractArray{T}...) where T
-@adjoint! function push!(xs::AbstractVector{<:AbstractArray}, x::AbstractArray{T}...) where T
-  sz_xs = size.(xs)
-  sz_x = size.(x)
-  push!(xs, x...), Δ -> begin
-    (Δ, map(x -> Ones{T}(x...), sz_x)...)
+# Exceptions:
+@adjoint! function push!(dst::AbstractVector, xs::AbstractArray...)
+  num_xs = length(xs)
+  push!(dst, xs...), Δ -> begin
+    (Δ[1:end-num_xs], Δ[end-num_xs+1:end]...)
   end
 end
 
-@adjoint! function pop!(xs::AbstractVector{<:AbstractArray{T}}) where T
-  sz_xs = size.(xs)
-  op = pop!(xs)
-  op, Δ -> begin
-    ([Ones{T}(sz...) for sz in sz_xs], )
-  end
+@adjoint! function pop!(src::AbstractVector{<:AbstractArray})
+  zs = fill(nothing, length(src)-1)
+  pop!(src), Δ -> (vcat(zs, [Δ]),)
```

Comment:
Use `push` for clarity sake. 
```suggestion
  pop!(src), Δ -> (push!(zs, Δ),)
```

Reply:
Tests will fail if you do this.

Reply:
Why?

Reply:
You can try it out pretty easily:
```
julia> let
        z = fill(nothing, 3)
        push!(z, [4, 5])
       end
ERROR: MethodError: Cannot `convert` an object of type Vector{Int64} to an object of type Nothing
```

Reply:
I'm sorry, `z` is not `Vector{Array}}` anyway. `zs` needs to have a union eltype at construction. Apologies if that wasn't clear.

Reply:
I can't read your mind. If you write something in the suggestion box, I take that as your suggestion.

More broadly I think a bit of focus on the big picture would be great. This is a super-minor style point, on a WIP PR, while there are major "can this work at all?" questions in play.